pydantic/_internal/_generate_schema.py

```diff
@@ -210,31 +219,171 @@ def modify_model_json_schema(
     return json_schema
 
 
+class ConfigWrapperStack:
+    """A stack of `ConfigWrapper` instances."""
+
+    def __init__(self, config_wrapper: ConfigWrapper):
+        self._config_wrapper_stack: list[ConfigWrapper] = [config_wrapper]
+
+    @property
+    def tail(self) -> ConfigWrapper:
+        return self._config_wrapper_stack[-1]
+
+    def push(self, config_wrapper: ConfigWrapper | ConfigDict | None) -> ContextManager[None]:
+        if config_wrapper is None:
+            return nullcontext()
+
+        if not isinstance(config_wrapper, ConfigWrapper):
+            config_wrapper = ConfigWrapper(config_wrapper, check=False)
+
+        @contextmanager
+        def _context_manager() -> Iterator[None]:
+            self._config_wrapper_stack.append(config_wrapper)
+            try:
+                yield
+            finally:
+                self._config_wrapper_stack.pop()
+
+        return _context_manager()
+
+
+JsonEncoders = Dict[Type[Any], Callable[[Any], Any]]
+
+
+def _add_custom_serialization_from_json_encoders(
```

Comment:
Does it make sense for this to be a method of `GenerateSchema`? I'm okay either way.

Reply:
Don’t think it matters too much happy to change it